base/reduce.jl

```diff
@@ -1374,3 +1374,50 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
     end
     return n
 end
+
+"""
+    startswith(target, prefix)
+
+Simultaneously iterate both arguments until they disagree or one of them is exhausted;
+if they disagree or `target` is exhausted first, return `false`; if `prefix` is exhausted
+before or at the same time as `target` and they never disagree, return `true`.
+
+Results are well-defined for ordered collections only.
+
+# Examples
+```jldoctest
+julia> startswith((1,2,3), (1,2))
+true
+```
+"""
+function startswith(target, prefix)
+    length(prefix) > length(target) && return false
+    for (a, b) ∈ zip(target, prefix)
+        isequal(a, b) || return false
+    end
+    true
```

Comment:
The docstring says `iterate`, but the implementation uses `length` which is not an iterable property. I think it is pretty easy to convert to just `iterate`. This version has a slightly simpler the implementation and matches the docstring description, but iterates `t` further than necessary:
```suggestion
    t, p = iterate(target), iterate(prefix)
    while !(p === nothing)
        t === nothing && return false
        isequal(t[1], p[1]) || return false
        t, p = iterate(target, t[2]), iterate(prefix, p[2])
    end
    return true
```
I think a better version is to only consume exactly `p` from `t`, so that stateful iterators also work, even though this makes it a bit more complex:
```suggestion
    p = iterate(prefix)
    p === nothing && return true
    t = iterate(target)
    while true
        t === nothing && return false
        isequal(t[1], p[1]) || return false
        p = iterate(target, p[2])
        p === nothing && return true
        t = iterate(target)
    end
    return true
```